File: serve.js
#!/usr/bin/env node
/**
 * Tiny local HTTP server for Dungeon Gleaner development.
 * Bypasses file:// CORS restrictions so audio, JSON fetches, etc. all work.
 *
 * Usage:
 *     node serve.js          # serves on http://localhost:8080
 *     node serve.js 9000     # serves on http://localhost:9000
 *
 * Then open http://localhost:8080 in your browser.
 */
const http = require('http');
const fs   = require('fs');
const path = require('path');

const PORT = parseInt(process.argv[2], 10) || 8080;
const ROOT = __dirname;

const MIME = {
  '.html': 'text/html',           '.css': 'text/css',
  '.js':   'application/javascript', '.mjs': 'application/javascript',
  '.json': 'application/json',    '.png': 'image/png',
  '.jpg':  'image/jpeg',          '.gif': 'image/gif',
  '.svg':  'image/svg+xml',       '.webp': 'image/webp',
  '.webm': 'audio/webm',          '.opus': 'audio/opus',
  '.ogg':  'audio/ogg',           '.mp3': 'audio/mpeg',
  '.wav':  'audio/wav',           '.wasm': 'application/wasm',
};

http.createServer(function (req, res) {
  var url = decodeURIComponent(req.url.split('?')[0]);
  if (url === '/') url = '/index.html';
  var file = path.join(ROOT, url);

  // Security: stay inside ROOT
  if (!file.startsWith(ROOT)) { res.writeHead(403); res.end(); return; }

  fs.stat(file, function (err, stat) {
    if (err || !stat.isFile()) { res.writeHead(404); res.end('Not found'); return; }
    var ext = path.extname(file).toLowerCase();
    res.writeHead(200, {
      'Content-Type': MIME[ext] || 'application/octet-stream',
      'Content-Length': stat.size,
      'Access-Control-Allow-Origin': '*',
      'Cache-Control': 'no-cache',
    });
    fs.createReadStream(file).pipe(res);
  });
}).listen(PORT, function () {
  console.log('Dungeon Gleaner dev server → http://localhost:' + PORT);
  console.log('Press Ctrl+C to stop.\n');
});

File: serve.py
#!/usr/bin/env python3
"""
Tiny local HTTP server for Dungeon Gleaner development.
Bypasses file:// CORS restrictions so audio, JSON fetches, etc. all work.

Usage:
    python3 serve.py          # serves on http://localhost:8080
    python3 serve.py 9000     # serves on http://localhost:9000

Then open http://localhost:8080 in your browser.
"""
import http.server
import sys
import os

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8080

# Serve from the directory this script lives in
os.chdir(os.path.dirname(os.path.abspath(__file__)))

class CORSHandler(http.server.SimpleHTTPRequestHandler):
    """Adds CORS headers and correct MIME types for game assets."""

    # Extend MIME map for game assets
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        '.webm': 'audio/webm',
        '.opus': 'audio/opus',
        '.ogg': 'audio/ogg',
        '.mp3': 'audio/mpeg',
        '.wav': 'audio/wav',
        '.json': 'application/json',
        '.js': 'application/javascript',
        '.mjs': 'application/javascript',
        '.wasm': 'application/wasm',
    }

    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Cache-Control', 'no-cache')
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(200)
        self.end_headers()

    def log_message(self, fmt, *args):
        # Quieter logging — only errors
        if args and '404' in str(args[1] if len(args) > 1 else ''):
            super().log_message(fmt, *args)

print(f'Dungeon Gleaner dev server → http://localhost:{PORT}')
print('Press Ctrl+C to stop.\n')
http.server.HTTPServer(('', PORT), CORSHandler).serve_forever()
